File: go/states.go
package main

var (
	st_goodbye State
	st_say     State
	st_run     State
	st_around  State
	st_never   State
	st_gonna   State
	st_tell    State
	st_a_lie   State
	st_down    State
	st_cry     State
	st_up      State
	st_you     State
	st_let     State
	st_give    State
	st_make    State
	st_and     State
	st_desert  State
	st_hurt    State
)

func initStates() {
	st_goodbye = State{
		NextList: []*State{&st_never},
		Text:     "goodbye",
	}

	st_say = State{
		NextList: []*State{&st_goodbye},
		Text:     "say",
	}

	st_run = State{
		NextList: []*State{&st_around},
		Text:     "run",
	}

	st_around = State{
		NextList: []*State{&st_and},
		Text:     "around",
	}

	st_never = State{
		NextList: []*State{&st_gonna},
		Text:     "Never",
	}

	st_gonna = State{
		NextList: []*State{&st_say, &st_run, &st_tell, &st_let, &st_give, &st_make},
		Text:     "gonna",
	}

	st_tell = State{
		NextList: []*State{&st_a_lie},
		Text:     "tell",
	}

	st_a_lie = State{
		NextList: []*State{&st_and},
		Text:     "a lie",
	}

	st_down = State{
		NextList: []*State{&st_never},
		Text:     "down",
	}

	st_cry = State{
		NextList: []*State{&st_never},
		Text:     "cry",
	}

	st_up = State{
		NextList: []*State{&st_never},
		Text:     "up",
	}

	st_you = State{
		NextList: []*State{&st_never, &st_up, &st_cry, &st_down},
		Text:     "you",
	}

	st_let = State{
		NextList: []*State{&st_you},
		Text:     "let",
	}

	st_give = State{
		NextList: []*State{&st_you},
		Text:     "give",
	}

	st_make = State{
		NextList: []*State{&st_you},
		Text:     "make",
	}

	st_and = State{
		NextList: []*State{&st_desert, &st_hurt},
		Text:     "and",
	}

	st_desert = State{
		NextList: []*State{&st_you},
		Text:     "desert",
	}

	st_hurt = State{
		NextList: []*State{&st_you},
		Text:     "hurt",
	}
}

File: csharp_rick/src/States.cs
namespace State{
	struct Returns
	{
		public Base Next;
		public string Text;
	}

	abstract class Base {
		public abstract Returns runState();
	}

	abstract class SimpleBase : Base{
		protected Returns runSimple(Type next, string text)
		{	/*
			Function is more complicated than it needs to be, the final return could be sufficient.
			However, c# warns about possible nulls, coming from CreateInstance.
			This could happen if:
			- runSimple is called with "null" explicitly
			- MultiBase random logic had a bug, or bad parameters
			both of which are unlikely. But just suppressing the error feels wrong.
			*/
			
			object? next_NullableObj = Activator.CreateInstance(next);
			// Cast later, since the cast would warn about possible null.
			Base next_Base = (Base)(next_NullableObj ?? new SNever());
			return new Returns()
			{
				Next = next_Base,
				Text = text
			};
		}
	}

	abstract class MultiBase : SimpleBase {
		protected Returns runMulti(Type[] nextList, string text)
		{
			Type result = nextList[
				Program.rnd.Next(nextList.Count())
			];
			return runSimple(result, text);
		}
	}

	//// All states ////

	class SGoodbye : SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SNever), "goodbye");
		}
	}

	class SSay: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SGoodbye), "say");
		}
	}

	class SRun: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SAround), "run");
		}
	}

	class SAround: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SAnd), "around");
		}
	}

	class SNever: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SGonna), "Never");
		}
	}

	class SGonna: MultiBase
	{
		public override Returns runState()
		{
			return runMulti(
				nextList: new Type[]{
					typeof(SSay),
					typeof(SRun),
					typeof(STell),
					typeof(SLet),
					typeof(SGive),
					typeof(SMake)
				}, 
				text: "gonna"
			);
		}
	}

	class STell: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SALie), "tell");
		}
	}

	class SALie: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SAnd), "a lie");
		}
	}

	class SDown: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SNever), "down");
		}
	}

	class SCry: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SNever), "cry");
		}
	}

	class SUp: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SNever), "up");
		}
	}

	class SYou: MultiBase
	{
		public override Returns runState()
		{
			return runMulti(
				nextList: new Type[]{
					typeof(SNever),
					typeof(SUp),
					typeof(SCry),
					typeof(SDown)
				}, 
				text: "you"
			);
		}
	}

	class SLet: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SYou), "let");
		}
	}

	class SGive: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SYou), "give");
		}
	}

	class SMake: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SYou), "make");
		}
	}

	class SAnd: MultiBase
	{
		public override Returns runState()
		{
			return runMulti(
				nextList: new Type[]{
					typeof(SDesert),
					typeof(SHurt)
				}, 
				text: "and"
			);
		}
	}

	class SDesert: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SYou), "desert");
		}
	}

	class SHurt: SimpleBase
	{
		public override Returns runState()
		{
			return runSimple(typeof(SYou), "hurt");
		}
	}
}
